1st Import Libraries and then the dataset

In [4]:
import pandas as pd
import numpy as np
!pip install black

ny_credits = pd.read_csv('https://raw.githubusercontent.com/hypoport/ep-dna-bewerberaufgabe/52645279ddc032955dabfaf43479da3e261b0709/sample.csv')
ny_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 175862 entries, 0 to 175861
Data columns (total 28 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   action_taken_name               175862 non-null  object 
 1   agency_name                     175862 non-null  object 
 2   applicant_ethnicity_name        175862 non-null  object 
 3   applicant_race_name_1           175862 non-null  object 
 4   applicant_sex_name              175862 non-null  object 
 5   applicant_income_000s           151344 non-null  float64
 6   co_applicant_ethnicity_name     175862 non-null  object 
 7   co_applicant_race_name_1        175862 non-null  object 
 8   co_applicant_sex_name           175862 non-null  object 
 9   county_name                     175377 non-null  object 
 10  denial_reason_name_1            23067 non-null   object 
 11  denial_reason_name_2            4650 non-null    object 
 12  denial_reason_na

Here I change parameters in order to be able to see the data according to my preferences

In [37]:
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 1000)
pd.set_option('display.float_format', lambda x: '%.2f' % x)

I will make a copy of the dataset not to re-write the original. ".copy" is also possible. I just prefer it this way!
And I add a lambda function to calculate the DTI (when possible) according to the data given.

In [38]:
ny_credits_1 = ny_credits.assign(DTI_nr = lambda x: x.hud_median_family_income / (x.applicant_income_000s*1000))
ny_credits_1

action_taken_name  \
0       Application denied by financial institution   
1                Application withdrawn by applicant   
2                                   Loan originated   
3                                   Loan originated   
4                                   Loan originated   
...                                             ...   
175857                              Loan originated   
175858                              Loan originated   
175859                              Loan originated   
175860                              Loan originated   
175861                              Loan originated   

                                        agency_name applicant_ethnicity_name  \
0              Consumer Financial Protection Bureau   Not Hispanic or Latino   
1       Department of Housing and Urban Development   Not Hispanic or Latino   
2              Consumer Financial Protection Bureau   Not Hispanic or Latino   
3              Consumer Financial Protection Bureau   Not Hispanic or Latino   
4              National Credit Union Administration   Not Hispanic or Latino   
...                                             ...                      ...   
175857         Consumer Financial Protection Bureau   Not Hispanic or Latino   
175858  Department of Housing and Urban Development   Not Hispanic or Latino   
175859  Department of Housing and Urban Development   Not Hispanic or Latino   
175860  Department of Housing and Urban Development   Not Hispanic or Latino   
175861         Consumer Financial Protection Bureau   Not Hispanic or Latino   

       applicant_race_name_1 applicant_sex_name  applicant_income_000s  \
0                      White             Female                 176.00   
1                      White               Male                  53.00   
2                      White               Male                 223.00   
3                      Asian             Female                  43.00   
4                      White               Male                  72.00   
...                      ...                ...                    ...   
175857                 White             Female                 118.00   
175858                 White             Female                  27.00   
175859                 White               Male                 215.00   
175860                 White               Male                 116.00   
175861                 Asian             Female                 116.00   

       co_applicant_ethnicity_name co_applicant_race_name_1  \
0                  No co-applicant          No co-applicant   
1                  No co-applicant          No co-applicant   
2           Not Hispanic or Latino                    White   
3                  No co-applicant          No co-applicant   
4                  No co-applicant          No co-applicant   
...                            ...                      ...   
175857             No co-applicant          No co-applicant   
175858             No co-applicant          No co-applicant   
175859      Not Hispanic or Latino                    White   
175860             No co-applicant          No co-applicant   
175861      Not Hispanic or Latino                    Asian   

       co_applicant_sex_name        county_name denial_reason_name_1  \
0            No co-applicant  Chautauqua County           Collateral   
1            No co-applicant        Erie County                  NaN   
2                     Female      Queens County                  NaN   
3            No co-applicant      Queens County                  NaN   
4            No co-applicant     Steuben County                  NaN   
...                      ...                ...                  ...   
175857       No co-applicant      Nassau County                  NaN   
175858       No co-applicant      Monroe County                  NaN   
175859                Female      Nassau County                  NaN   
175860       No co-applicant    Rockland Count

Here I am transforming categorical values into booleans so that I can sum them later when I need to.

In [39]:
def transform_denial_name(df):
    df['Collateral'] = df["denial_reason_name_1"].str.match(r'Collateral')
    df['Debt-to-income ratio'] = df["denial_reason_name_1"].str.match(r'Debt-to-income ratio')
    df['Credit history'] = df["denial_reason_name_1"].str.match(r'Credit history')
    df['Employment history'] = df["denial_reason_name_1"].str.match(r'Employment history')
    df['Other'] = df["denial_reason_name_1"].str.match(r'Other')
    df['Insufficient cash (downpayment, closing costs)'] = df["denial_reason_name_1"].str.match(r'Insufficient cash (downpayment, closing costs)')
    df['Credit application incomplete'] = df["denial_reason_name_1"].str.match(r'Credit application incomplete')
    df['Unverifiable information'] = df["denial_reason_name_1"].str.match(r'Unverifiable information')
    df['Mortgage insurance denied'] = df["denial_reason_name_1"].str.match(r'Mortgage insurance denied')
    return df

denial_to_num_df = transform_denial_name(ny_credits_1)
denial_to_num_df.head(50)

action_taken_name  \
0   Application denied by financial institution   
1            Application withdrawn by applicant   
2                               Loan originated   
3                               Loan originated   
4                               Loan originated   
5             Loan purchased by the institution   
6             Loan purchased by the institution   
7                               Loan originated   
8                               Loan originated   
9   Application denied by financial institution   
10  Application denied by financial institution   
11  Application denied by financial institution   
12                              Loan originated   
13                              Loan originated   
14        Application approved but not accepted   
15  Application denied by financial institution   
16                              Loan originated   
17                              Loan originated   
18  Application denied by financial institution   
19                              Loan originated   
20            Loan purchased by the institution   
21                              Loan originated   
22                              Loan originated   
23               File closed for incompleteness   
24                              Loan originated   
25                              Loan originated   
26                              Loan originated   
27                              Loan originated   
28            Loan purchased by the institution   
29           Application withdrawn by applicant   
30                              Loan originated   
31  Application denied by financial institution   
32                              Loan originated   
33            Loan purchased by the institution   
34  Application denied by financial institution   
35                              Loan originated   
36                              Loan originated   
37                              Loan originated   
38                              Loan originated   
39  Application denied by financial institution   
40        Application approved but not accepted   
41                              Loan originated   
42                              Loan originated   
43            Loan purchased by the institution   
44                              Loan originated   
45                              Loan originated   
46                              Loan originated   
47                              Loan originated   
48  Application denied by financial institution   
49                              Loan originated   

                                    agency_name  \
0          Consumer Financial Protection Bureau   
1   Department of Housing and Urban Development   
2          Consumer Financial Protection Bureau   
3          Consumer Financial Protection Bureau   
4          National Credit Union Administration   
5          Consumer Financial Protection Bureau   
6          Consumer Financial Protection Bureau   
7   Department of Housing and Urban Development   
8   Department of Housing and Urban Development   
9   Department of Housing and Urban Development   
10         Consumer Financial Protection Bureau   
11         National Credit Union Administration   
12         National Credit Union Administration   
13         National Credit Union Administration   
14  Department of Housing and Urban Development   
15         Consumer Financial Protection Bureau   
16         Consumer Financial Protection Bureau   
17         National Credit Union Administration   
18  Department of Housing and Urban Development   
19         Consumer Financial Protection Bureau   
20  Department of Housing and Urban Development   
21  Department of Housing and Urban Development   
22         Consumer Financial Protection Bureau   
23         Consumer Financial Protection Bureau   
24         National Credit Union Administration   
25         Consumer Financial Protection Bureau   
26         National Credit Union Administration   


In [40]:
denial_to_num_df.to_csv('reasons_loan_denied.csv')

Take an overview of your nummerical columns including the calculation of the DTI 
which is a creteria to deny loans if DTI > 43%

In [41]:
actions_in_numbers = (
    ny_credits_1[["action_taken_name","applicant_ethnicity_name", "applicant_income_000s", "hud_median_family_income", "loan_amount_000s", "DTI_nr", 'agency_name', 'denial_reason_name_1', 'denial_reason_name_2', 'denial_reason_name_3', 'lien_status_name', 'loan_purpose_name', 'loan_type_name']]
    .sort_values(by='DTI_nr', ascending= False)
)
actions_in_numbers.head(300)

action_taken_name  \
171188  Application denied by financial institution   
103786  Application denied by financial institution   
28553            Application withdrawn by applicant   
45706   Application denied by financial institution   
10018   Application denied by financial institution   
96997                File closed for incompleteness   
14209   Application denied by financial institution   
28219            Application withdrawn by applicant   
77478   Application denied by financial institution   
142316  Application denied by financial institution   
5333                                Loan originated   
61580   Application denied by financial institution   
124731                              Loan originated   
173021           Application withdrawn by applicant   
10736         Application approved but not accepted   
99836            Application withdrawn by applicant   
142172  Application denied by financial institution   
53043   Application denied by financial institution   
111995  Application denied by financial institution   
97000                               Loan originated   
50523                               Loan originated   
74845   Application denied by financial institution   
17059                               Loan originated   
59913   Application denied by financial institution   
150274  Application denied by financial institution   
26824   Application denied by financial institution   
40705                File closed for incompleteness   
102300  Application denied by financial institution   
40071   Application denied by financial institution   
84632   Application denied by financial institution   
166563  Application denied by financial institution   
99530   Application denied by financial institution   
24291   Application denied by financial institution   
160341  Application denied by financial institution   
126205           Application withdrawn by applicant   
21970   Application denied by financial institution   
121717  Application denied by financial institution   
55651   Application denied by financial institution   
47717            Application withdrawn by applicant   
120754  Application denied by financial institution   
85637   Application denied by financial institution   
127587  Application denied by financial institution   
122021  Application denied by financial institution   
38853   Application denied by financial institution   
35100   Application denied by financial institution   
107340           Application withdrawn by applicant   
43503   Application denied by financial institution   
120187  Application denied by financial institution   
55511                               Loan originated   
26833   Application denied by financial institution   
7001    Application denied by financial institution   
38252   Application denied by financial institution   
113563  Application denied by financial institution   
77033   Application denied by financial institution   
87821            Application withdrawn by applicant   
111470  Application denied by financial institution   
55811   Application denied by financial institution   
141722  Application denied by financial institution   
88813                               Loan originated   
166777               File closed for incompleteness   
98836   Application denied by financial institution   
148924        Application approved but not accepted   
72713            Application withdrawn by applicant   
33967                               Loan originated   
161766  Application denied by financial institution   
130304                              Loan originated   
85615   Application denied by financial institution   
77751   Application denied by financial institution   
98772   Application denied by financial institution   
108349  Application denied by financial institution   
27838            Application withdrawn by applicant   
155447               File closed for incompleteness   
63485              

In [42]:
actions_in_numbers.to_csv('searching_for_patterns_in_numbers.csv')

You will find many credits accepted inspite of score > 43%

This could be explained by either the type of credit or the credit institution, which loan fees may be cheaper




Here I created 2 functions to be able to filter data here and in the future.

In [43]:
def filter_field(df, f1, v1, f2, v2):
    return df[(f1 == v1) & (f2 == v2)]

def filter_field2(df, f1, v1, f2, v2):
    return df[(f1 == v1) & (f2 != v2)]

In [44]:
#Testing function
filter_field(actions_in_numbers, actions_in_numbers.action_taken_name, "Application denied by financial institution", actions_in_numbers.agency_name, "Office of the Comptroller of the Currency").count()


action_taken_name           2373
applicant_ethnicity_name    2373
applicant_income_000s       2258
hud_median_family_income    2373
loan_amount_000s            2373
DTI_nr                      2258
agency_name                 2373
denial_reason_name_1        2373
denial_reason_name_2         544
denial_reason_name_3         106
lien_status_name            2373
loan_purpose_name           2373
loan_type_name              2373
dtype: int64

In the following step I am replacing the values on the column "actions_taken" for "Credit" when the application was accepted, and "No_Credit" when the loan was denied.

Here it's not relevant whether the client took the credit or not in the end, because I am focusing on the reasons why the credit institution had approved or denied the applications. 

So I will drop those values. Besides, those categories ('Loan purchased by the institution'and 'Application withdrawn by applicant') have a lot of Nans, so it will also help with the cleaning!

In [45]:
#action_reasons = pd.Series(['Application denied by financial institution', 'File closed for incompleteness',
#                            'Preapproval request denied by financial institution', np.nan]).str.replace('No Credit','No Credit','No Credit', regex=True, inplace = True)
action_reasons= actions_in_numbers
action_reasons["action_taken_name"].replace(
    {"Application denied by financial institution": "No_credit", "File closed for incompleteness": "No_credit", 
     'Preapproval request denied by financial institution': "No_credit", "Loan originated": "Credit", 
     "Application approved but not accepted": "Credit"},
    inplace=True)


In [46]:
#To drop rows that contain any value in the list I 1st define values:
values_actions = ['Loan purchased by the institution', 'Application withdrawn by applicant']

#Now I can drop the rows containing any value in the defined list
action_reasons = action_reasons[action_reasons.action_taken_name.isin(values_actions) == False]

action_reasons.head()

action_taken_name                           applicant_ethnicity_name  \
171188         No_credit  Information not provided by applicant in mail,...   
103786         No_credit                             Not Hispanic or Latino   
45706          No_credit                             Not Hispanic or Latino   
10018          No_credit                             Not Hispanic or Latino   
96997          No_credit                             Not Hispanic or Latino   

        applicant_income_000s  hud_median_family_income  loan_amount_000s  \
171188                   1.00                 109000.00                10   
103786                   1.00                 109000.00               626   
45706                    1.00                 109000.00                10   
10018                    1.00                 109000.00                10   
96997                    1.00                 109000.00                20   

        DTI_nr                                agency_name  \
171188  109.00       Consumer Financial Protection Bureau   
103786  109.00  Office of the Comptroller of the Currency   
45706   109.00       Consumer Financial Protection Bureau   
10018   109.00       Consumer Financial Protection Bureau   
96997   109.00       National Credit Union Administration   

            denial_reason_name_1 denial_reason_name_2 denial_reason_name_3  \
171188      Debt-to-income ratio                  NaN                  NaN   
103786      Debt-to-income ratio                  NaN                  NaN   
45706             Credit history                  NaN                  NaN   
10018   Unverifiable information                  NaN                  NaN   
96997                        NaN                  NaN                  NaN   

                     lien_status_name loan_purpose_name loan_type_name  
171188          Not secured by a lien  Home improvement   Conventional  
103786        Secured by a first lien     Home purchase   Conventional  
45706   Secured by a subordinate lien  Home improvement   Conventional  
10018           Not secured by a lien  Home improvement   Conventional  
96997         Secured by a first lien  Home improvement   Conventional

From the data I filtered, 71.5% of the cases the application has been accepted, and 28.5% has been denied

In [47]:
action_reasons.action_taken_name.str.count('No_credit').sum()

38690

In [48]:
action_reasons.action_taken_name.str.count('Credit').sum()


96725

# LET'S TAKE A LOOK AT THE DENIED CASES

In [49]:
No_credit = action_reasons.loc[action_reasons['action_taken_name'] == 'No_credit'].sort_values(by= "DTI_nr", ascending = True)
No_credit.drop(columns=['applicant_income_000s', 'hud_median_family_income'])
No_credit.head(300)

action_taken_name                           applicant_ethnicity_name  \
86791          No_credit  Information not provided by applicant in mail,...   
125682         No_credit                             Not Hispanic or Latino   
137688         No_credit                             Not Hispanic or Latino   
26324          No_credit  Information not provided by applicant in mail,...   
48582          No_credit  Information not provided by applicant in mail,...   
30314          No_credit                             Not Hispanic or Latino   
76729          No_credit                             Not Hispanic or Latino   
54144          No_credit                             Not Hispanic or Latino   
108728         No_credit  Information not provided by applicant in mail,...   
55414          No_credit                             Not Hispanic or Latino   
169881         No_credit                             Not Hispanic or Latino   
19941          No_credit                             Not Hispanic or Latino   
79932          No_credit                             Not Hispanic or Latino   
12443          No_credit                             Not Hispanic or Latino   
19182          No_credit  Information not provided by applicant in mail,...   
127418         No_credit                             Not Hispanic or Latino   
89008          No_credit                             Not Hispanic or Latino   
38343          No_credit                             Not Hispanic or Latino   
162191         No_credit                             Not Hispanic or Latino   
17074          No_credit  Information not provided by applicant in mail,...   
106487         No_credit  Information not provided by applicant in mail,...   
56682          No_credit                             Not Hispanic or Latino   
98395          No_credit                             Not Hispanic or Latino   
164268         No_credit  Information not provided by applicant in mail,...   
106709         No_credit                             Not Hispanic or Latino   
86797          No_credit                             Not Hispanic or Latino   
32610          No_credit                             Not Hispanic or Latino   
135275         No_credit  Information not provided by applicant in mail,...   
137749         No_credit  Information not provided by applicant in mail,...   
12215          No_credit                             Not Hispanic or Latino   
65277          No_credit                             Not Hispanic or Latino   
108206         No_credit  Information not provided by applicant in mail,...   
134466         No_credit                             Not Hispanic or Latino   
164508         No_credit  Information not provided by applicant in mail,...   
57430          No_credit                             Not Hispanic or Latino   
66954          No_credit                                 Hispanic or Latino   
140631         No_credit                             Not Hispanic or Latino   
148290         No_credit                             Not Hispanic or Latino   
118971         No_credit                             Not Hispanic or Latino   
16535          No_credit                             Not Hispanic or Latino   
70593          No_credit  Information not provided by applicant in mail,...   
49075          No_credit                             Not Hispanic or Latino   
158362         No_credit                             Not Hispanic or Latino   
12396          No_credit                             Not Hispanic or Latino   
135709         No_credit                             Not Hispanic or Latino   
44474          No_credit  Information not provided by applicant in mail,...   
162244         No_credit                             Not Hispanic or Latino   
48482          No_credit  Information not provided by applicant in mail,...   
131849         No_credit                             Not Hispanic or Latino   
109565         No_credit                             Not 

In [50]:
No_credit.to_csv('searching_for_No_credit_reasons.csv')

40% of the "No_credit" has no clear reason for denial. Only a 7% from that (1124 applications) do not count with a Secured lien, and could be explained by it.

In [51]:
Denial_unclear= No_credit.loc[No_credit['denial_reason_name_1'].isnull()].sort_values(by= "DTI_nr", ascending = False)
Denial_unclear


action_taken_name                           applicant_ethnicity_name  \
96997          No_credit                             Not Hispanic or Latino   
55651          No_credit                                 Hispanic or Latino   
99530          No_credit  Information not provided by applicant in mail,...   
40705          No_credit                             Not Hispanic or Latino   
142172         No_credit                             Not Hispanic or Latino   
...                  ...                                                ...   
175138         No_credit                             Not Hispanic or Latino   
175341         No_credit                                 Hispanic or Latino   
175417         No_credit                             Not Hispanic or Latino   
175541         No_credit                             Not Hispanic or Latino   
175662         No_credit                             Not Hispanic or Latino   

        applicant_income_000s  hud_median_family_income  loan_amount_000s  \
96997                    1.00                 109000.00                20   
55651                    1.00                  71300.00               413   
99530                    1.00                  71300.00               700   
40705                    1.00                  71300.00              2800   
142172                   1.00                  71300.00               203   
...                       ...                       ...               ...   
175138                    NaN                 109000.00               469   
175341                    NaN                  71300.00               436   
175417                    NaN                  71300.00               165   
175541                    NaN                  71300.00               323   
175662                    NaN                  95200.00               365   

        DTI_nr                                  agency_name  \
96997   109.00         National Credit Union Administration   
55651    71.30  Department of Housing and Urban Development   
99530    71.30  Department of Housing and Urban Development   
40705    71.30         Consumer Financial Protection Bureau   
142172   71.30  Department of Housing and Urban Development   
...        ...                                          ...   
175138     NaN  Department of Housing and Urban Development   
175341     NaN  Department of Housing and Urban Development   
175417     NaN  Department of Housing and Urban Development   
175541     NaN  Department of Housing and Urban Development   
175662     NaN  Department of Housing and Urban Development   

       denial_reason_name_1 denial_reason_name_2 denial_reason_name_3  \
96997                   NaN                  NaN                  NaN   
55651                   NaN                  NaN                  NaN   
99530                   NaN                  NaN                  NaN   
40705                   NaN                  NaN                  NaN   
142172                  NaN                  NaN                  NaN   
...                     ...                  ...                  ...   
175138                  NaN                  NaN                  NaN   
175341                  NaN                  NaN                  NaN   
175417                  NaN                  NaN                  NaN   
175541                  NaN                  NaN                  NaN   
175662                  NaN                  NaN                  NaN   

               lien_status_name loan_purpose_name loan_type_name  
96997   Secured by a first lien  Home improvement   Conventional  
55651   Secured by a first lien       Refinancing   Conventional  
99530   Secured by a first lien     Home purchase   Conventional  
40705   Secured by a first lien  Home improvement   Conventional  
142172  Secured by a first lien       Refinancing   Conventional  
...                         ...               ...            ...  
175138  Secured by a first lie

The Etnicity has called my attention in the past. Is is there an obvious tendency of denial due to this?
Let's use the filter I created befofre:

In [52]:
filter_field(Denial_unclear, Denial_unclear.applicant_ethnicity_name, "Hispanic or Latino", Denial_unclear.action_taken_name, "No_credit").count()

action_taken_name           1165
applicant_ethnicity_name    1165
applicant_income_000s        978
hud_median_family_income    1153
loan_amount_000s            1165
DTI_nr                       966
agency_name                 1165
denial_reason_name_1           0
denial_reason_name_2           3
denial_reason_name_3          16
lien_status_name            1165
loan_purpose_name           1165
loan_type_name              1165
dtype: int64

Only 7% (1165 from 15623 unclear denial) are Latinos, this is not enough evidence for me to continue looking at that field.

No Credit Reasons: 

11% or more applications that have not been accepted do not count with a "lien"
To be more sure about this correlation a "correlation coefficient" could be calculated
This does not explain for certain though that there is causality

In [53]:
No_credit.lien_status_name.str.count('Not secured by a lien').sum()


4313

And another 7% doesnt have a lien

In [54]:
Denial_unclear.lien_status_name.str.count('Not secured by a lien').sum()

1124

# HERE WE FOCUSE ON APPROVED CREDITS

In [55]:
Credit = action_reasons.loc[action_reasons['action_taken_name'] == 'Credit']
Credit.drop(columns=['applicant_income_000s', 'applicant_ethnicity_name', 'hud_median_family_income', 'denial_reason_name_1', 'denial_reason_name_2', 'denial_reason_name_3'])

action_taken_name  loan_amount_000s  DTI_nr  \
5333              Credit               362  109.00   
124731            Credit               184   82.70   
10736             Credit               200   71.30   
97000             Credit              2647   71.30   
50523             Credit               276   71.30   
...                  ...               ...     ...   
175694            Credit               359     NaN   
175748            Credit              1165     NaN   
175767            Credit               485     NaN   
175794            Credit               515     NaN   
175811            Credit                98     NaN   

                                        agency_name  \
5333    Department of Housing and Urban Development   
124731  Department of Housing and Urban Development   
10736   Department of Housing and Urban Development   
97000     Office of the Comptroller of the Currency   
50523          Consumer Financial Protection Bureau   
...                                             ...   
175694  Department of Housing and Urban Development   
175748         Consumer Financial Protection Bureau   
175767         Consumer Financial Protection Bureau   
175794  Department of Housing and Urban Development   
175811        Federal Deposit Insurance Corporation   

                     lien_status_name loan_purpose_name loan_type_name  
5333          Secured by a first lien       Refinancing    FHA-insured  
124731        Secured by a first lien     Home purchase   Conventional  
10736         Secured by a first lien     Home purchase   Conventional  
97000         Secured by a first lien       Refinancing   Conventional  
50523         Secured by a first lien       Refinancing   Conventional  
...                               ...               ...            ...  
175694        Secured by a first lien       Refinancing   Conventional  
175748  Secured by a subordinate lien  Home improvement   Conventional  
175767        Secured by a first lien       Refinancing   Conventional  
175794        Secured by a first lien       Refinancing    FHA-insured  
175811        Secured by a first lien     Home purchase   Conventional  

[96725 rows x 7 columns]

In [56]:
Credit.to_csv('searching_for_patterns_in_approved_credits.csv')

94% of the approved credits count with a "Secured lien"

In [57]:
Credit.lien_status_name.str.count('Secured by a first lien').sum()

88427

In [58]:
Credit.lien_status_name.str.count('Secured by a subordinate lien').sum()


3408

6% of the approved credits do not have a lien

In [59]:
Credit.lien_status_name.str.count('Not secured by a lien').sum()

4890

In [60]:
#What about the DTI??
Credit.loc[Credit['DTI_nr'] >43].count()

action_taken_name           7
applicant_ethnicity_name    7
applicant_income_000s       7
hud_median_family_income    7
loan_amount_000s            7
DTI_nr                      7
agency_name                 7
denial_reason_name_1        0
denial_reason_name_2        0
denial_reason_name_3        0
lien_status_name            7
loan_purpose_name           7
loan_type_name              7
dtype: int64

Here I am creating a new dataframe to evaluate the specific cases that were approved despite of not having a lien.
Can I identify other factors???

In [61]:
No_lien = Credit.loc[Credit['lien_status_name'] == 'Not secured by a lien']
No_lien.drop(columns=['applicant_income_000s', 'hud_median_family_income', 'denial_reason_name_1', 'denial_reason_name_2', 'denial_reason_name_3'])

action_taken_name                           applicant_ethnicity_name  \
55511             Credit                             Not Hispanic or Latino   
63485             Credit  Information not provided by applicant in mail,...   
54892             Credit                                 Hispanic or Latino   
159122            Credit                             Not Hispanic or Latino   
98820             Credit                             Not Hispanic or Latino   
...                  ...                                                ...   
161233            Credit                             Not Hispanic or Latino   
161674            Credit                             Not Hispanic or Latino   
168776            Credit                             Not Hispanic or Latino   
171098            Credit                             Not Hispanic or Latino   
174783            Credit                             Not Hispanic or Latino   

        loan_amount_000s  DTI_nr                            agency_name  \
55511                  4   63.00   National Credit Union Administration   
63485                  5   34.85   Consumer Financial Protection Bureau   
54892                  5   23.77   Consumer Financial Protection Bureau   
159122                 7   23.23   Consumer Financial Protection Bureau   
98820                  4   23.23   Consumer Financial Protection Bureau   
...                  ...     ...                                    ...   
161233                 9     NaN                 Federal Reserve System   
161674                10     NaN  Federal Deposit Insurance Corporation   
168776                 5     NaN   Consumer Financial Protection Bureau   
171098                15     NaN   Consumer Financial Protection Bureau   
174783               500     NaN                 Federal Reserve System   

             lien_status_name loan_purpose_name loan_type_name  
55511   Not secured by a lien  Home improvement   Conventional  
63485   Not secured by a lien  Home improvement   Conventional  
54892   Not secured by a lien  Home improvement   Conventional  
159122  Not secured by a lien  Home improvement   Conventional  
98820   Not secured by a lien  Home improvement   Conventional  
...                       ...               ...            ...  
161233  Not secured by a lien  Home improvement   Conventional  
161674  Not secured by a lien  Home improvement   Conventional  
168776  Not secured by a lien  Home improvement   Conventional  
171098  Not secured by a lien  Home improvement   Conventional  
174783  Not secured by a lien  Home improvement   Conventional  

[4890 rows x 8 columns]

Here I am running my filter to quickly see if "Agency Name" could give me some insights on the above task

In [62]:
filter_field(No_lien, No_lien.agency_name, "National Credit Union Administration", No_lien.loan_type_name, "Conventional").count()

action_taken_name           1854
applicant_ethnicity_name    1854
applicant_income_000s       1849
hud_median_family_income    1845
loan_amount_000s            1854
DTI_nr                      1840
agency_name                 1854
denial_reason_name_1           0
denial_reason_name_2           0
denial_reason_name_3           0
lien_status_name            1854
loan_purpose_name           1854
loan_type_name              1854
dtype: int64

The Agency Name returns interesting values to continue the analysis on the field

Here I am transforming the categorical values into booleans to be able to sum them later

In [63]:
def transform_action_name(df):
    df['action_loan_denied'] = df["action_taken_name"].str.match(r'^Application\sdenied')
    df['action_application_withdrawn'] = df["action_taken_name"].str.match(r'^Application\swithdrawn')
    df['action_loan_originated'] = df["action_taken_name"].str.match(r'^Loan\soriginated')
    df['action_loan_purchased'] = df["action_taken_name"].str.match(r'^Loan\spurchased')
    df['action_loan_not_taken'] = df["action_taken_name"].str.match(r'^Application\sapproved.*but not accepted')  
    df['action_file_incomplete'] = df["action_taken_name"].str.match(r'File closed for incompletenes')
    df['action_preapproval_denied'] = df["action_taken_name"].str.match(r'Preapproval request denied by financial institution')

    return df

enriched_df = transform_action_name(ny_credits_1)

Here I want to get an overview of the actions taken grouped by agency

In [64]:
Credit_by_agency = (
    enriched_df[["agency_name", "action_loan_denied", "action_application_withdrawn", "action_loan_originated",
                 "action_loan_purchased", "action_loan_not_taken", "action_file_incomplete", "action_preapproval_denied"]]
    .groupby(["agency_name"]).sum())

Credit_by_agency

action_loan_denied  \
agency_name                                                       
Consumer Financial Protection Bureau                      13564   
Department of Housing and Urban Development               11245   
Federal Deposit Insurance Corporation                       872   
Federal Reserve System                                      754   
National Credit Union Administration                       3197   
Office of the Comptroller of the Currency                  2373   

                                             action_application_withdrawn  \
agency_name                                                                 
Consumer Financial Protection Bureau                                 5431   
Department of Housing and Urban Development                          6713   
Federal Deposit Insurance Corporation                                 589   
Federal Reserve System                                                297   
National Credit Union Administration                                 1995   
Office of the Comptroller of the Currency                             740   

                                             action_loan_originated  \
agency_name                                                           
Consumer Financial Protection Bureau                          30626   
Department of Housing and Urban Development                   30960   
Federal Deposit Insurance Corporation                          4070   
Federal Reserve System                                         2680   
National Credit Union Administration                          14048   
Office of the Comptroller of the Currency                      8598   

                                             action_loan_purchased  \
agency_name                                                          
Consumer Financial Protection Bureau                         16618   
Department of Housing and Urban Development                   6622   
Federal Deposit Insurance Corporation                           23   
Federal Reserve System                                          29   
National Credit Union Administration                             9   
Office of the Comptroller of the Currency                     1381   

                                             action_loan_not_taken  \
agency_name                                                          
Consumer Financial Protection Bureau                          1501   
Department of Housing and Urban Development                   2547   
Federal Deposit Insurance Corporation                          337   
Federal Reserve System                                         200   
National Credit Union Administration                           587   
Office of the Comptroller of the Currency                      571   

                                             action_file_incomplete  \
agency_name                                                           
Consumer Financial Protection Bureau                           3108   
Department of Housing and Urban Development                    2345   
Federal Deposit Insurance Corporation                           283   
Federal Reserve System                                          142   
National Credit Union Administration                            543   
Office of the Comptroller of the Currency                       261   

                                             action_preapproval_denied  
agency_name                                                             
Consumer Financial Protection Bureau                                 2  
Department of Housing and Urban Development                          0  
Federal Deposit Insurance Corporation                                0  
Federal Reserve System                                               0  
National Credit Union Administration                                 1  
Office of the Comptroller of the Currency                            0

Let's see the loan's average amount per agency.
Further analysis can be done in terms of: 
    Which company gave more credits, and which one gives more money per credit.
    This can give an insight towards what type of interest/motivation each institution may have to give a loan

In [65]:
Loans_by_agency = (
    ny_credits_1[["agency_name", "loan_amount_000s"]]
    .groupby(["agency_name"])
    .agg({"loan_amount_000s": ["sum", "count"]})
)
Loans_by_agency.assign(ratio=lambda r: r[('loan_amount_000s', 'sum')]/r[('loan_amount_000s', 'count')]).sort_values(by="ratio")

loan_amount_000s         ratio
                                                         sum  count       
agency_name                                                               
National Credit Union Administration                 2963754  20380 145.42
Office of the Comptroller of the Currency            3485847  13924 250.35
Department of Housing and Urban Development         17151998  60432 283.82
Federal Reserve System                               1218084   4102 296.95
Federal Deposit Insurance Corporation                2303391   6174 373.08
Consumer Financial Protection Bureau                31489580  70850 444.45

# This analysis is a bit more isolated and falls on the limit of the time i had to complete the task. This is why I didn't go deeper into this topic.
Nevertheless, I would like to have a very brief look here at the applicants' sex.
Which sex has applied the most? Can we find any tendency in the resolution based on this creteria?

In [66]:
def transform_sex_name(df):
    df['Female'] = df["applicant_sex_name"].str.match(r'Female')
    df['Male'] = df["applicant_sex_name"].str.match(r'Male')

    return df

ny_credits_1 = transform_sex_name(ny_credits_1)

#loan decision by sex
loans_by_sex = (
    ny_credits_1[["action_taken_name", "Female", "Male"]]
    .groupby(["action_taken_name"]).sum()
)
loans_by_sex

Female   Male
action_taken_name                                                
Application approved but not accepted                 1586   3663
Application denied by financial institution          10232  18592
Application withdrawn by applicant                    4559   9610
File closed for incompleteness                        1834   4075
Loan originated                                      27415  56031
Loan purchased by the institution                     2404   4965
Preapproval request denied by financial institu...       0      3

In [67]:
loans_by_sex.to_csv('loans_by_sex.csv')

# THE END

In [68]:
# Supporting research not used:
enriched_df[enriched_df.applicant_ethnicity_name == 'Hispanic or Latino']['applicant_ethnicity_name'].count()

10015

There are 10015 applicants who are latinos